### ISSUE TITLE
[TinyYolo v2] Bug in maxpool?

### ISSUE THREAD

**Issue by owulveryck** [owulveryck/onnx-go#73]
[This commit](https://github.com/owulveryck/onnx-go/commit/f864ca8f8e05a1d21170073b8f63cc6793ed0ac6) allows the model [tiny Yolo v2](https://github.com/onnx/models/tree/master/tiny_yolov2) to be compiled and executed With Gorgonia.

Sadly the execution does not give the expected result:

```
➜  model_zoo_executor git:(tiny-yolov2) ✗  export MODELDIR=~/Documents/tiny_yolov2
➜  model_zoo_executor git:(tiny-yolov2) ✗ go run main.go -model $MODELDIR/model.onnx -input $MODELDIR/test_data_set_0/input_0.pb -output $MODELDIR/test_data_set_0/output_0.pb

        Error Trace:    main.go:72
                                                proc.go:200
                                                asm_amd64.s:1337
        Error:          Max difference between -0.17929432 and 0.056231752 allowed is 0.005, but difference was -0.23552606999874115
        Messages:       the two tensors should be equal.
exit status 1
```

According to this [blog post](http://machinethink.net/blog/object-detection-with-yolo/) the architecture should be:
```
Layer         kernel  stride  output shape
---------------------------------------------
Input                          (416, 416, 3)
Convolution    3×3      1      (416, 416, 16)
MaxPooling     2×2      2      (208, 208, 16)
Convolution    3×3      1      (208, 208, 32)
MaxPooling     2×2      2      (104, 104, 32)
Convolution    3×3      1      (104, 104, 64)
MaxPooling     2×2      2      (52, 52, 64)
Convolution    3×3      1      (52, 52, 128)
MaxPooling     2×2      2      (26, 26, 128)
Convolution    3×3      1      (26, 26, 256)
MaxPooling     2×2      2      (13, 13, 256)
Convolution    3×3      1      (13, 13, 512)
MaxPooling     2×2      1      (13, 13, 512)
Convolution    3×3      1      (13, 13, 1024)
Convolution    3×3      1      (13, 13, 1024)
Convolution    1×1      1      (13, 13, 125)
---------------------------------------------
```

After setting some logs, the architecture of the decoded network is:
```diff
+Convolution             (3, 3)          [1 1]           (1, 16, 416, 416)
+MaxPooling              (2, 2)          [2 2]           (1, 16, 208, 208)
+Convolution             (3, 3)          [1 1]           (1, 32, 208, 208)
+MaxPooling              (2, 2)          [2 2]           (1, 32, 104, 104)
+Convolution             (3, 3)          [1 1]           (1, 64, 104, 104)
+MaxPooling              (2, 2)          [2 2]           (1, 64, 52, 52)
+Convolution             (3, 3)          [1 1]           (1, 128, 52, 52)
+MaxPooling              (2, 2)          [2 2]           (1, 128, 26, 26)
+Convolution             (3, 3)          [1 1]           (1, 256, 26, 26)
+MaxPooling              (2, 2)          [2 2]           (1, 256, 13, 13)
+Convolution             (3, 3)          [1 1]           (1, 512, 13, 13)
-MaxPooling              (2, 2)          [1 1]           (1, 512, 14, 14)
-Convolution             (3, 3)          [1 1]           (1, 1024, 14, 14)
-Convolution             (3, 3)          [1 1]           (1, 1024, 14, 14)
-Convolution             (1, 1)          [1 1]           (1, 125, 14, 14)
```

The last layer using the `Maxpool` operator does not give the correct output size.
The padding used is computed from the `auto_pad` argument but seems ok (padding is `[1,1]`).

It requires more investigation; maybe a bug in Gorgonia.

_Note_ : the computation is slow, but `Make it work`, then `Make it fast`

cc @chewxy

**Comment by owulveryck** [comment]
Maybe related: [PR 274](https://github.com/gorgonia/gorgonia/pull/274) from Gorgonia (from @lynic)

**Comment by chewxy** [comment]
Does  yolo use overlapping pooling? i.e. maxpool with a stride of other than size

**Comment by owulveryck** [comment]
@chewxy I don't think so; the stride of the operator is `[1 1]`

**Comment by owulveryck** [comment]
I think I got it; the ONNX spec specifies that a `ceil_mode` attribute can change the behavior of the shape calculus:

> ceil_mode : int (default is 0)
> Whether to use ceil or floor (default) to compute the output shape.

The [`calcShape`](https://github.com/gorgonia/gorgonia/blob/fd7a1bd48ae0654826c28375fd63553ad7505f22/op_nn.go#L764) method in Gorgonia does not take this into account; the shape calculation is probably different than what onnx expects.

I will do different tests, but If this is the case, I will have to find a smart solution to implement this without breaking Gorgonia's API 🧐

**Comment by owulveryck** [comment]
I did not notice that the implementation of the `SAME_UPPER` was breaking the unit tests of gorgonnx:

```
--- FAIL: TestONNX (0.04s)
    --- FAIL: TestONNX/TestMaxpool2dSameUpper (0.00s)
        test_structure.go:78:
                Error Trace:    test_structure.go:135
                Error:          Max difference between 1.7640524 and 0.978738 allowed is 1e-06, but difference was 0.7853143811225891
                Messages:       the two tensors should be equal.
    --- FAIL: TestONNX/TestMaxpool2dPrecomputedSameUpper (0.00s)
        test_structure.go:78:
                Error Trace:    test_structure.go:135
                Error:          Max difference between 1 and 7 allowed is 1e-06, but difference was -6
                Messages:       the two tensors should be equal.
FAIL
```

Let's fix this before going further.

**Comment by owulveryck** [comment]
Probably related to [ONNX issue #1113](https://github.com/onnx/onnx/issues/1113)

**Comment by owulveryck** [comment]
The problem linked to #74 and because the padding is asymmetric;
this is not implemented in Gorgonia.

**Comment by owulveryck** [comment]
Should be fixed with PR #80 and [PR #295 from Gorgonia](https://github.com/gorgonia/gorgonia/pull/295)

**Comment by owulveryck** [comment]
Tiny yolo v2 is working, let's make it fast now!
